skipruntime-ts/ts/examples/database.ts

```diff
@@ -0,0 +1,173 @@
+import type {
+  SKStore,
+  TJSON,
+  Mapper,
+  EagerCollection,
+  NonEmptyIterator,
+  SimpleSkipService,
+  SimpleServiceOutput,
+  JSONObject,
+  Writer,
+} from "skip-runtime";
+
+import { runWithServer } from "skip-runtime";
+
+import sqlite3 from "sqlite3";
+
+/*****************************************************************************/
+// Populate the database with made-up values (if it's not already there)
+/*****************************************************************************/
+
+async function initDB(): Promise<sqlite3.Database> {
+  const db = new sqlite3.Database("./db.sqlite");
+  // Create the table if it doesn't exist
+  await db.exec(`CREATE TABLE IF NOT EXISTS data (
+    id TEXT PRIMARY KEY,
+    object JSON
+    )`);
+
+  await db.run(
+    "INSERT OR REPLACE INTO data (id, object) VALUES ($id, $object)",
+    {
+      $id: "123",
+      $object: JSON.stringify({
+        name: "daniel",
+        country: "FR",
+      }),
+    },
+  );
+  await db.run(
+    "INSERT OR REPLACE INTO data (id, object) VALUES ($id, $object)",
+    {
+      $id: "124",
+      $object: JSON.stringify({
+        name: "josh",
+        country: "UK",
+      }),
+    },
+  );
+  await db.run(
+    "INSERT OR REPLACE INTO data (id, object) VALUES ($id, $object)",
+    {
+      $id: "125",
+      $object: JSON.stringify({
+        name: "julien",
+        country: "ES",
+      }),
+    },
+  );
+  return db;
+}
+
+const db = await initDB();
+
+/*****************************************************************************/
+// The protocol
+/*****************************************************************************/
+
+type Command = {
+  command: "getUser" | "set" | "delete";
+  payload: TJSON;
+};
+
+type GetUser = string;
+type Set = { key: string; value: string }[];
+type Delete = { keys: string[] }[];
+
+/*****************************************************************************/
+// The read path, we want to find a user
+/*****************************************************************************/
+
+type User = {
+  name: string;
+  country: string;
+};
+
+type Response = { status: "error"; msg: string } | { status: "ok"; user: User };
+
+class Request implements Mapper<string, Command, string, Response> {
+  constructor(private users: EagerCollection<string, TJSON>) {}
+
+  mapElement(
+    key: string,
+    it: NonEmptyIterator<Command>,
+  ): Iterable<[string, Response]> {
+    const cmd = it.first();
+    const value = this.users.maybeGetOne(cmd.payload as GetUser);
+    const user = value as User;
+    const computed: Response =
+      cmd.command == "getUser"
+        ? { status: "ok", user }
+        : { status: "error", msg: "Unknown command" };
+    return Array([key, computed]);
+  }
+}
```

Comment:
Bit simpler with ValueMapper, since you're not modifying key structure at all here.

```suggestion
class Request implements ValueMapper<string, Command, Response> {
  constructor(private users: EagerCollection<string, TJSON>) {}

  mapValue(
    cmd : Command,
    key: string,
  ): Response {
    const value = this.users.maybeGetOne(cmd.payload as GetUser);
    const user = value as User;
    return cmd.command == "getUser"
        ? { status: "ok", user }
        : { status: "error", msg: "Unknown command" };
  }
}
```